### ISSUE TITLE
Experimenting with an unsupported model, Polaris 9650iq

### ISSUE THREAD

**Issue by VynDesign** [galletn/iaqualink#20]
I figured I'd give it a try, what the heck. The attributes shown are accurately returning my account information, so the authentication is working, but I'm getting the "Device does not belong to user" status reported:

<img width="437" alt="Screenshot 2024-06-27 170212" src="https://github.com/galletn/iaqualink/assets/882135/676e39ec-15a0-4504-b1b0-d9ebf3d34ae9">

I'm willing to do any legwork you would like for me to assist with getting this model supported. If you have a basic postman collection I can use to start investigating my specific model, please share it.

**Comment by galletn** [comment]
Hi there!

the good news is, it's not the first time we get this error back, mostly it was because multiple devices exist on the account, and for example some of the equipment returns null for owner id, when you then request further data it returns this error.

Now let's hope it's something like that, and not that you need yet another API, then we will need to use a proxy to trace what api's are called.

attached the postman collection, login, then get devices, and please post that reply here so that I can have a first look at what happens. 

[New Collection Test.postman_collection.json](https://github.com/user-attachments/files/16025552/New.Collection.Test.postman_collection.json)

**Comment by VynDesign** [comment]
I figured out which values from the 'login' response were required for subsequent calls. It was a little confusing which token was needed by the device /shadow, /features, and /site endpoints as the names varied and didn't exactly match what was in the login response - turns out it was the userPoolOAuth.IdToken value. I modified the postman collection to use collectionVariables to store the values necessary, and then the subsequent calls reference the collectionVariables:

[iAquaLink Robot API template.postman_collection.json](https://github.com/user-attachments/files/16031853/iAquaLink.Robot.API.template.postman_collection.json)

You just need to input the username and password for your specific account, and the rest get populated when you hit 'login' and then 'devices':

<img width="643" alt="Screenshot 2024-06-28 101433" src="https://github.com/galletn/iaqualink/assets/882135/6f418c78-6a8e-4a08-8c3a-99ee9c00db20">

I can get the list of devices (I only have the one):

<img width="586" alt="Screenshot 2024-06-28 103326" src="https://github.com/galletn/iaqualink/assets/882135/73ddbd97-48a6-4426-bbf7-0f2fb4c5aaf4">

Unfortunately, when I attempt to get the /shadow, /features, or /site, I get the "Device does not belong to user" response (in various formats):

![Screenshot 2024-06-28 103656](https://github.com/galletn/iaqualink/assets/882135/91b9ceb4-9359-45d8-9bb7-c4671f3cc0e6)

<img width="576" alt="Screenshot 2024-06-28 103214" src="https://github.com/galletn/iaqualink/assets/882135/4c337a41-9fc7-40a3-88db-2c019bf5baec">

![Screenshot 2024-06-28 103956](https://github.com/galletn/iaqualink/assets/882135/8acbdb96-469b-44b3-a3c1-2df7a8757527)

**Comment by galletn** [comment]
Good news is I found some time to check out the websocket thing, seems to be fairly easy.

Can you try and run following Python code: (replace with your credentials and device ID

import asyncio
import json
import datetime
import requests
import websocket
from websockets.sync.client import connect

def run():
        attributes = {}
        headers = {"Content-Type": "application/json; charset=utf-8", "Connection": "keep-alive", "Accept": "*/*" }
        url = "https://prod.zodiac-io.com/users/v1/login"
        data = {"apikey": "EOOEMOW4YR6QNB07", "email": "$mail$", "password": "$Pass$"}
        data = json.dumps(data)
        response = requests.post(url, headers = headers, data = data)
        print(response)

        if response.status_code == 200:
            data = response.json()
            first_name = data["first_name"]
            last_name = data["last_name"]
            authentication_token = data["authentication_token"]
            id =  data["id"]
            id_token = data["userPoolOAuth"]["IdToken"]

        print(id_token)

        message = { "action": "subscribe", "namespace": "authorization", "payload": { "userId": id }, "service": "Authorization", "target": "$Device Serial Number$", "version": 1 }
        y = json.dumps(message)

        with connect("wss://prod-socket.zodiac-io.com/devices", additional_headers={"Authorization": id_token}) as websocket:
            websocket.send(y)
            message = websocket.recv()
            
        print(message)

run()

**Comment by VynDesign** [comment]
It gets the JWT just fine, but then hangs until it closes the connection:

```Traceback (most recent call last):
  File "c:\Users\vynsa\Dev\python\iAquaLink\main.py", line 41, in <module>
    run()
  File "c:\Users\vynsa\Dev\python\iAquaLink\main.py", line 37, in run
    message = websocket.recv()
              ^^^^^^^^^^^^^^^^
  File "C:\Users\vynsa\AppData\Local\Programs\Python\Python311\Lib\site-packages\websockets\sync\connection.py", line 201, in recv
    raise self.protocol.close_exc from self.recv_events_exc
websockets.exceptions.ConnectionClosedOK: received 1001 (going away) Going away; then sent 1001 (going away) Going away
```
I also tried this in postman, no response is received:

<img width="636" alt="Screenshot 2024-06-28 152115" src="https://github.com/galletn/iaqualink/assets/882135/6508fe89-da89-4955-812d-9d51ae6949a7">

**Comment by VynDesign** [comment]
As for mitmproxy I have it installed, but don't really get how to use it just yet. I can see traffic from my laptop, but not other devices on the network from what I can tell. I was also using wireshark, but haven't seen any traffic to/from my vacuum's IP address. If you have any pointers, I'm all ears.

**Comment by falinka** [comment]
> It gets the JWT just fine, but then hangs until it closes the connection:
> 
> ```
>   File "c:\Users\vynsa\Dev\python\iAquaLink\main.py", line 41, in <module>
>     run()
>   File "c:\Users\vynsa\Dev\python\iAquaLink\main.py", line 37, in run
>     message = websocket.recv()
>               ^^^^^^^^^^^^^^^^
>   File "C:\Users\vynsa\AppData\Local\Programs\Python\Python311\Lib\site-packages\websockets\sync\connection.py", line 201, in recv
>     raise self.protocol.close_exc from self.recv_events_exc
> websockets.exceptions.ConnectionClosedOK: received 1001 (going away) Going away; then sent 1001 (going away) Going away
> ```
> 
> I also tried this in postman, no response is received:
> 
> <img alt="Screenshot 2024-06-28 152115" width="636" src="https://private-user-images.githubusercontent.com/882135/344296199-6508fe89-da89-4955-812d-9d51ae6949a7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk3Mzc3MDgsIm5iZiI6MTcxOTczNzQwOCwicGF0aCI6Ii84ODIxMzUvMzQ0Mjk2MTk5LTY1MDhmZTg5LWRhODktNDk1NS04MTJkLTlkNTFhZTY5NDlhNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYzMFQwODUwMDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMjUzMDA0MjVkNWNmZDhlMjAzY2Q0ODVhYjk2OThhY2NlYTNmZjQ4ZTc3ODg4MTZmMDM1OGQ1NjQwYjVmY2JiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.GD8D8tvUKV26wqlMPb4WKX0F5GdJ_QKq67X8JbtT3b0">

Hi VynDesign,
I think i had the similar issue, timing out on the same place forever. Can you try to manually put the serial number of your device in the script, instead of having it as variable ("target": "$Device Serial Number$" replace with "target": "Q9Zxxxxx")?

**Comment by VynDesign** [comment]
Yeah, I had already done that. And, unfortunately, now I won't be able to test for a while, as my robot is out of commission. One of the wheel motors went bad, so I'm waiting on a new motor block :(

**Comment by galletn** [comment]
too bad, ping me once you have it back, or we can check with what was still of last values... its not that important that it is online. if you are willing send me over your credentials (after you changed your pass first) then I can have a look. You can send them to nicolas.gallet@skynet.be

**Comment by falinka** [comment]
Is the robot out of game, or just you cannot start the cleaning mode? I believe plenty of valuable data can be gathered via MITM, in order to get modes, set modes :) Anyway, good luck with replacement, hope you don't for a long time.

**Comment by RogueDruid** [comment]
> The motor is replaced, but I'm getting Err10 on the control unit now 😒. I have a few other pieces that need replacing, which is probably causing slipping gears at the wheels. We're gearing up for our 4th of July party, so won't be able to do much until the weekend at this point.

err 10 is communication error. Most likely due to a cable getting pulled out of place where it connects to the control board. If you open up the connector there are screws to tighten the terminals.

**Comment by galletn** [comment]
hi @VynDesign the good news is that I found what is wrong for you, its the same is #4 you have an i2d_robot, and that model uses other endpoints/api's. The owner of the other thread has shared me his credentials and will put the robot in the pool next week, so I should be able to debug what needs to be done to get i2d_robot model supported too! Hope it will all be done by the time you have yours back!

**Comment by galletn** [comment]
@VynDesign https://github.com/galletn/iaqualink/tree/galletn-patch-4 contains the required code for your robot, can you test it?

**Comment by galletn** [comment]
should be fine now!

**Comment by VynDesign** [comment]
Sorry to be out of communication, life gets in the way sometimes. I never got a chance to purchase the rest of the parts I needed, so the robot's been sitting dormant. It's nearing the end of the pool season, but maybe I'll get a chance to fix it and test soon...
